src/main/java/org/zalando/fahrschein/NakadiReader.java

```diff
@@ -240,22 +249,30 @@ void runInternal(long timeout, TimeUnit timeoutUnit) throws IOException, Backoff
 
                 LOG.debug("Cursor for [{}] partition [{}] at offset [{}]", eventName, cursor.getPartition(), cursor.getOffset());
 
+                metricsCollector.ifPresent(mc -> mc.markBatchesReceived());
+
                 final JsonToken token = jsonParser.nextToken();
                 if (token != JsonToken.END_OBJECT) {
                     expectField(jsonParser, "events");
 
                     final List<T> events = readEvents(jsonParser);
+                    metricsCollector.ifPresent(mc -> mc.markEventsReceived(events.size()));
 
                     expectToken(jsonParser, JsonToken.END_OBJECT);
 
                     final Batch<T> batch = new Batch<>(cursor, Collections.unmodifiableList(events));
 
                     processBatch(batch);
+                } else {
+                    // it's a keep alive batch
+                    metricsCollector.ifPresent(mc -> mc.markEventsReceived(0));
```

Comment:
I counted them as events received because then there is no lack of data if we only receive keep-alive batches for a certain amount of time. Furthermore, you can see on your dashboard that there are really no events inside the messages and everything works fine.

True, I skipped the "eventsSuccessfullyProcessed" meter because I considered it to be monitored by the listener if it could process events or not. Moreover, we are handing over a batch of events to the listener and therefore do not know how many events of that batch were successfully processed and how many were erroneous. We can only tell about the entire batch (message, respectively) whether it was processed or not. We can have a "messagesSuccessfullyProcessed" as you proposed?

> messages received (instead of batches received), this includes keep-alive messages

"messagesReceived" is also fine for me, I change it.

> events processed successfully

See above

> messages failed

You mean renaming "errorWhileConsuming" to "messagesFailed"? I named it "errorWhileConsuming" because you don't really know at this point whether it was an exception in processing a batch or an error in the connection to Nakadi.

> maybe messages processed (messages received should then equal messages failed + messages processed)

I will add this (see above).